src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
+            tail = head;
+        } else if (index == size) {
+            add(value);
+            return;
+        } else {
+            current = getNodeByIndex(index);
+            current = new Node<>(current.prev, value, current);
+            current.next.prev = current;
+            current.prev.next = current;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T node : list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        current = getNodeByIndex(index);
+        T oldObject = current.item;
+        current.item = value;
+        return oldObject;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        current = getNodeByIndex(index);
+        if (current.next == null && current.prev == null) {
+            head = null;
+            tail = null;
+        } else if (current.next == null) {
+            current.prev.next = null;
+            tail = current.prev;
+        } else if (current.prev == null) {
+            current.next.prev = null;
+            head = current.next;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
+        current = head;
+        for (int i = 0; i < size; i++) {
+            if ((object == current.item)
+                    || (object != null && object.equals(current.item))) {
+                remove(i);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == EMPTY_LIST_VALUE;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndexExistence(index);
+        boolean isIndexLessHalf = index <= size / DEFAULT_DIVISOR;
+        Node<T> node;
+        if (isIndexLessHalf) {
+            node = head;
+            while (index != EMPTY_LIST_VALUE) {
+                node = node.next;
+                index--;
+            }
+        } else {
+            node = tail;
+            int stepCounter = size - index;
+            while (stepCounter != INDEX_STOPPING_FROM_TAIL) {
+                node = node.prev;
+                stepCounter--;
```

Comment:
I think we can make it simplier.
Maybe we can avoid working with two different variables. stepCounter and index.